Mackenzie/include alternate modalities

mackenziesnyder opened the pull request:
Completed:
- added the t2w modality as an option for autoafids
- this includes the synthSR program to convert t2w nifti files to t1w nifti files

Still To do:
- verify correct test data is used - SynthSR does not produce a high quality t1w image from the t2w image upon initial visual inspection of the files

Dhananjhay commented:
@mackenziesnyder Can you please update the DAG in the `README` with the new rule `synthsr`, thanks!

mackenziesnyder commented:
> @mackenziesnyder Can you please update the DAG in the `README` with the new rule `synthsr`, thanks!

@Dhananjhay I added the DAG in the latest commit

Dhananjhay commented:
Perfect, looks great! 

> Still To do:
> 
>  verify correct test data is used - SynthSR does not produce a high quality t1w image from the t2w image upon initial visual inspection of the files

Once you are done with this remaining task I think we'd be in a good shape to merge the branch. @ataha24 What do you think?

ataha24 commented on autoafids/config/snakebids.yml:
```diff
@@ -89,7 +89,7 @@ parse_args:
     
   --modality: 
-      help: 'Specify the template to use (e.g., T1w or T2w).'
+      help: 'Specify the modality to use (e.g., T1w or T2w).'
```
I would say either t1w or 'alternative' (SynthSR conversion). This is mainly because SynthSR would support any MRI!

mackenziesnyder replied:
Ok I think 'alternative' makes the most sense - I think I have to do a few other changes because in rule SynthSR I hard coded the suffix to look for t2w since I previously thought we needed a template for each and we only had t1w and t2w. I can look into making this generic instead now that I know t1w is the only template we need!

ataha24 replied:
Here are some related qs that arise that we can think about together. Probably best to dicuss more on a quick call! 

- What should the behaviour for this rule be when there are multiple files BUT no T1w image (e.g., T2w, Flair, or even CT)? 
- What if the dataset is heterogenous (i.e., some subjects have T1w others have T2w). Is that something we want to handle or even possible to handle? 

I am taking a look at some of the SynthSR outputs and I think they are indeed in the same "space" but there are some local mophological differences.

ataha24 commented:
> Perfect, looks great!
> 
> > Still To do:
> > verify correct test data is used - SynthSR does not produce a high quality t1w image from the t2w image upon initial visual inspection of the files
> 
> Once you are done with this remaining task I think we'd be in a good shape to merge the branch. @ataha24 What do you think?

I am happy with this so far as it is the best we can do to support other modelities. Ok to merge this, just left a small comment regarding how to invoke SynthSR during a run

Dhananjhay commented:
I'm getting an error when I try running autoafids on T2w modality

Command:
```
autoafids data/ test-data participant --participant-label 719222 --cores all --modality T2w
```

Error log:
```
Config file config/snakebids.yml is extended by additional config specified via the command line.
..... fast preprocessing configration .....
 the folowing squence will be triggered ['n4_bias_correction', 'norm_im', 'resample_im']
KeyError in file /local/scratch/autoafids/autoafids/workflow/rules/cnn.smk, line 29:
'T1w'
  File "/local/scratch/autoafids/autoafids/workflow/Snakefile", line 267, in <module>
  File "/local/scratch/autoafids/autoafids/workflow/rules/cnn.smk", line 29, in <module>
  File "/local/scratch/autoafids/.venv/lib/python3.10/site-packages/snakebids/core/datasets.py", line 627, in __getitem__
```

The dataset contains both T1w and T2w contrast images.

mackenziesnyder commented:
> I'm getting an error when I try running autoafids on T2w modality
> 
> Command:
> 
> ```
> autoafids data/ test-data participant --participant-label 719222 --cores all --modality T2w
> ```
> 
> Error log:
> 
> ```
> Config file config/snakebids.yml is extended by additional config specified via the command line.
> ..... fast preprocessing configration .....
>  the folowing squence will be triggered ['n4_bias_correction', 'norm_im', 'resample_im']
> KeyError in file /local/scratch/autoafids/autoafids/workflow/rules/cnn.smk, line 29:
> 'T1w'
>   File "/local/scratch/autoafids/autoafids/workflow/Snakefile", line 267, in <module>
>   File "/local/scratch/autoafids/autoafids/workflow/rules/cnn.smk", line 29, in <module>
>   File "/local/scratch/autoafids/.venv/lib/python3.10/site-packages/snakebids/core/datasets.py", line 627, in __getitem__
> ```
> 
> It might be because the pipeline is bypassing running `synthsr` rule first before looking for T1w image.

Ah this is because I specified to only load the pybids_inputs of the user specified modality - just going to change 'T1w' to config['modality'] in cnn.smk and I believe that should fix this. It works for dry runs so I'm just gonna test and then push!

mackenziesnyder commented:
I added functionality for T2w and FLAIR as per synthsr compatibility. 
I also added a warning message for when t2w or flair are chosen - indicates to the user that the predictions are only as good as sythsr's conversion from those modalities to a t1w image

ataha24 commented:
I tested `--modality T2w` on 20 MRIs. It ran perfectly fine for some subjects but failed on others because of memory. This may be something we should control at the synthsr rule level? I think its hard for the workflow to interpret how much memory is needed when parallelizing across subjects because of calling singularity so it is probably trying to fit as many subjects as there are cpus then later running out of memory.  Below is image to help with that: 

<img width="1703" alt="Screenshot 2025-02-19 at 2 57 17 PM" src="https://github.com/user-attachments/assets/1240b3cc-2da4-4488-b696-ce78503993fa" />

Otherwise, the workflow ran perfectly. One other minor comment is re how synthSR floods the terminal with outputs. Is there a nice way to gently suppressing that to make it user friendly to parse? Makes sense if that output is dumped into some log file (it may be already) but it may be too much to have everything be on terminal. What do you guys think?

<img width="1710" alt="Screenshot 2025-02-19 at 2 54 56 PM" src="https://github.com/user-attachments/assets/8982cf71-e1f4-4630-8694-9099b32c61fe" />

Summary: 
Issue A: need to allocate some memory for the rule so it does not overuse memory. 
Enhancement A: limit terminal outputs so user is not overwhelmed with outputs. 

Big picture questions: 
1) are we using the appropriate synthSR model? 
2) can sythSR support ct images and if so, could be cool feature to add later

mackenziesnyder commented:
> I tested `--modality T2w` on 20 MRIs (using --cores all). It ran perfectly fine for some subjects but failed on others because of memory. This may be something we should control at the synthsr rule level? I think its hard for the workflow to interpret how much memory is needed when parallelizing across subjects because of calling singularity so it is probably trying to fit as many subjects as there are cpus then later running out of memory. Below is image to help with that:
> 
> <img alt="Screenshot 2025-02-19 at 2 57 17 PM" width="1703" src="https://private-user-images.githubusercontent.com/46094728/414881699-1240b3cc-2da4-4488-b696-ce78503993fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5OTYzOTksIm5iZiI6MTczOTk5NjA5OSwicGF0aCI6Ii80NjA5NDcyOC80MTQ4ODE2OTktMTI0MGIzY2MtMmRhNC00NDg4LWI2OTYtY2U3ODUwMzk5M2ZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE5VDIwMTQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkMzQwNjA2OGQ5MmQ5NmU4ZWFkMDZlM2E4NjQyNDllOGY1ZDY4NDE1YzI1ZDZiODJjYTFmY2FiNjc5NTFhMTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HcstAJhPeqc60O3Hkqv6iCQLXn1NHqz2oblQMrIFmeU">
> Otherwise, the workflow ran perfectly. One other minor comment is re how synthSR floods the terminal with outputs. Is there a nice way to gently suppressing that to make it user friendly to parse? Makes sense if that output is dumped into some log file (it may be already) but it may be too much to have everything be on terminal. What do you guys think?
> 
> <img alt="Screenshot 2025-02-19 at 2 54 56 PM" width="1710" src="https://private-user-images.githubusercontent.com/46094728/414880756-8982cf71-e1f4-4630-8694-9099b32c61fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5OTYzOTksIm5iZiI6MTczOTk5NjA5OSwicGF0aCI6Ii80NjA5NDcyOC80MTQ4ODA3NTYtODk4MmNmNzEtZTFmNC00NjMwLTg2OTQtOTA5OWIzMmM2MWZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE5VDIwMTQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwOGEwNTc5NWM2MDc5YzRmNGE1YjlhZGQ0ZTViY2QzYWNkNTgwZGI5YjViNDFjMDNiMDg5ZWJlMmNhZTRlNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Vew8_16eFL9OM0XupIrbWdxLhxXbhPIS5dtmjOrtDeY">
> Summary: Issue A: need to allocate some memory for the rule so it does not overuse memory. Enhancement A: limit terminal outputs so user is not overwhelmed with outputs.
> 
> Big picture questions:
> 
> 1. are we using the appropriate synthSR model?
> 2. can sythSR support ct images and if so, could be cool feature to add later

Hi Alaa, I will try and test the workflow on a larger dataset to see if I can reproduce the error. I'll also look into putting the terminal output in a log file. 

to answer the big picture questions:

mackenziesnyder commented:
@ataha24 I have done a lot of testing today - Synthsr uses a TON of memory when it runs (over 59% of my memory at its highest when only running one subject) so I believe it would have significant issues when trying to paralyze it. I can specify in the rule that it needs to be run independently and cannot be paralyzed through resource allocations. I used cbs basic and it would crash pretty much immediately when trying to run more than 1 subject at a time - I'm not sure if you were running it on cbs heavy previously to have some subjects run ok.

Dhananjhay commented:
Hmm I have never seen this error before:

Command:
```
autoafids myelin/ test-myelein participant --participant-label 718216 --cores all --modality T2w --use-singularity
```

Error:
```
Traceback (most recent call last):
  File "/local/scratch/autoafids/.venv/bin/auto_afids_cnn_apply", line 6, in <module>
    sys.exit(main())
  File "/local/scratch/autoafids/autoafids/workflow/scripts/apply.py", line 486, in main
    predictions = apply_all(args.model, img,args.sub_prior)
  File "/local/scratch/autoafids/autoafids/workflow/scripts/apply.py", line 350, in apply_all
    model = tf.keras.models.load_model(model_architecture_path)
  File "/local/scratch/autoafids/.venv/lib/python3.10/site-packages/keras/src/saving/saving_api.py", line 206, in load_model
    raise ValueError(
ValueError: File format not supported: filepath=/tmp/tmpj20kl3az/afid-01.model. Keras 3 only supports V3 `.keras` files and legacy H5 format files (`.h5` extension). Note that the legacy SavedModel format is not supported by `load_model()` in Keras 3. In order to reload a TensorFlow SavedModel as an inference-only layer in Keras 3, use `keras.layers.TFSMLayer(/tmp/tmpj20kl3az/afid-01.model, call_endpoint='serving_default')` (note that your `call_endpoint` might have a different name).
[Fri Feb 21 00:58:30 2025]
Error in rule gen_fcsv:
    jobid: 1
    input: work/sub-718216/ses-01/resample/sub-718216_ses-01_res-100_desc-zscore_T1w.nii.gz, /localscratch/.cache/autoafids/model/model-combined.afidsmodel, work/sub-718216/ses-01/registration/sub-718216_ses-01_space-native_desc-MNI_afids.fcsv
    output: ./sub-718216/ses-01/afids-cnn/sub-718216_ses-01_desc-afidscnn_afids.fcsv
    log: logs/sub-718216/ses-01/sub-718216_ses-01_landmark.log (check log file(s) for error details)
    shell:
        auto_afids_cnn_apply work/sub-718216/ses-01/resample/sub-718216_ses-01_res-100_desc-zscore_T1w.nii.gz /localscratch/.cache/autoafids/model/model-combined.afidsmodel ./sub-718216/ses-01/afids-cnn/sub-718216_ses-01_desc-afidscnn_afids.fcsv work/sub-718216/ses-01/registration/sub-718216_ses-01_space-native_desc-MNI_afids.fcsv
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)
```

ataha24 commented:
> Hmm I have never seen this error before:
> 
> Command:
> 
> ```
> autoafids myelin/ test-myelein participant --participant-label 718216 --cores all --modality T2w --use-singularity
> ```
> 
> Error:
> 
> ```
> Traceback (most recent call last):
>   File "/local/scratch/autoafids/.venv/bin/auto_afids_cnn_apply", line 6, in <module>
>     sys.exit(main())
>   File "/local/scratch/autoafids/autoafids/workflow/scripts/apply.py", line 486, in main
>     predictions = apply_all(args.model, img,args.sub_prior)
>   File "/local/scratch/autoafids/autoafids/workflow/scripts/apply.py", line 350, in apply_all
>     model = tf.keras.models.load_model(model_architecture_path)
>   File "/local/scratch/autoafids/.venv/lib/python3.10/site-packages/keras/src/saving/saving_api.py", line 206, in load_model
>     raise ValueError(
> ValueError: File format not supported: filepath=/tmp/tmpj20kl3az/afid-01.model. Keras 3 only supports V3 `.keras` files and legacy H5 format files (`.h5` extension). Note that the legacy SavedModel format is not supported by `load_model()` in Keras 3. In order to reload a TensorFlow SavedModel as an inference-only layer in Keras 3, use `keras.layers.TFSMLayer(/tmp/tmpj20kl3az/afid-01.model, call_endpoint='serving_default')` (note that your `call_endpoint` might have a different name).
> [Fri Feb 21 00:58:30 2025]
> Error in rule gen_fcsv:
>     jobid: 1
>     input: work/sub-718216/ses-01/resample/sub-718216_ses-01_res-100_desc-zscore_T1w.nii.gz, /localscratch/.cache/autoafids/model/model-combined.afidsmodel, work/sub-718216/ses-01/registration/sub-718216_ses-01_space-native_desc-MNI_afids.fcsv
>     output: ./sub-718216/ses-01/afids-cnn/sub-718216_ses-01_desc-afidscnn_afids.fcsv
>     log: logs/sub-718216/ses-01/sub-718216_ses-01_landmark.log (check log file(s) for error details)
>     shell:
>         auto_afids_cnn_apply work/sub-718216/ses-01/resample/sub-718216_ses-01_res-100_desc-zscore_T1w.nii.gz /localscratch/.cache/autoafids/model/model-combined.afidsmodel ./sub-718216/ses-01/afids-cnn/sub-718216_ses-01_desc-afidscnn_afids.fcsv work/sub-718216/ses-01/registration/sub-718216_ses-01_space-native_desc-MNI_afids.fcsv
>         (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)
> ```

Hey Djay, error has to do with using a recent tensorflow version which stopped supporting our models! Is there a reason why it is defaulting to that version?

Dhananjhay commented:
I've rebased `mackenzie/include_alternate_modalities` onto djay/docker - this required solving merge conflicts but in long run it gives us a linear history of commits and eventually would be easier when we merge this PR. Hey @mackenziesnyder please let me know if you see witness any odd looking lines of code or any weird behaviour!

Dhananjhay commented:
Hey @ataha24 I think it was because we earlier had this logic in the pyproject.toml:
```
tensorflow = [
    { version="<=2.16.0", python = "<3.11" },
    { version=">=2.12.0, <=2.16.0", python = ">=3.11" }
]
```
which was downloading `tensorflow v2.16.0rc0`. After rebase and poetry installing `autoafids` again, `tensorflow` was downgraded to `v2.11.1` and I was able to run this branch successfully!

Dhananjhay commented:
> @ataha24 I have done a lot of testing today - Synthsr uses a TON of memory when it runs (over 59% of my memory at its highest when only running one subject) so I believe it would have significant issues when trying to paralyze it. I can specify in the rule that it needs to be run independently and cannot be paralyzed through resource allocations. I tried memory allocation for the rule so it doesn't overuse memory, but it was still consistently crashing unfortunately. I used cbs basic and it would crash pretty much immediately when trying to run more than 1 subject at a time - I'm not sure if you were running it on cbs heavy previously to have some subjects run ok.

I was able to run 27 subjects in parallel by allocating 32000 MB of memory and 4 threads for `synthsr` rule on CBS heavy which I believe can now be extrapolated to as many subjects you want. But like @mackenziesnyder said, it fails almost instantly on CBS basic. This will require some further investigation.

mackenziesnyder commented:
> > @ataha24 I have done a lot of testing today - Synthsr uses a TON of memory when it runs (over 59% of my memory at its highest when only running one subject) so I believe it would have significant issues when trying to paralyze it. I can specify in the rule that it needs to be run independently and cannot be paralyzed through resource allocations. I tried memory allocation for the rule so it doesn't overuse memory, but it was still consistently crashing unfortunately. I used cbs basic and it would crash pretty much immediately when trying to run more than 1 subject at a time - I'm not sure if you were running it on cbs heavy previously to have some subjects run ok.
> 
> I was able to run 27 subjects in parallel by allocating 32000 MB of memory and 4 threads for `synthsr` rule on CBS heavy which I believe can now be extrapolated to as many subjects you want. But like @mackenziesnyder said, it fails almost instantly on CBS basic. This will require some further investigation.

I looked into this a bit more because I noticed Synthsr was used in another program of Ali's I'm currently working with (snakedwi) - it is specified in the code to use 8 threads for the rule. I tried that implementation and it results in it not being paralyzed on cbs basic as well which is unfortunate - it defaults to the 4 threads that are available and will only run one subject at a time.

ataha24 commented:
> > > @ataha24 I have done a lot of testing today - Synthsr uses a TON of memory when it runs (over 59% of my memory at its highest when only running one subject) so I believe it would have significant issues when trying to paralyze it. I can specify in the rule that it needs to be run independently and cannot be paralyzed through resource allocations. I tried memory allocation for the rule so it doesn't overuse memory, but it was still consistently crashing unfortunately. I used cbs basic and it would crash pretty much immediately when trying to run more than 1 subject at a time - I'm not sure if you were running it on cbs heavy previously to have some subjects run ok.
> > 
> > 
> > I was able to run 27 subjects in parallel by allocating 32000 MB of memory and 4 threads for `synthsr` rule on CBS heavy which I believe can now be extrapolated to as many subjects you want. But like @mackenziesnyder said, it fails almost instantly on CBS basic. This will require some further investigation.
> 
> I looked into this a bit more because I noticed Synthsr was used in another program of Ali's I'm currently working with (snakedwi) - it is specified in the code to use 8 threads for the rule. I tried that implementation and it results in it not being paralyzed on cbs basic as well which is unfortunate - it defaults to the 4 threads that are available and will only run one subject at a time.

This is good! I think most SynthSR users are aware its resource intensive. I think we did well so long as the workflow runs successfully. From a big data perspective, users would probably be fine b/c typically they would run on a compute cluster!

ataha24 commented:
Are you all happy to merge this 😎?

Dhananjhay commented:
Yes, push your changes and we can test them out on both CBS basic and heavy.

Dhananjhay commented:
@ataha24 @mackenziesnyder I believe we're in good shape to merge this branch. I've added some additional logic, building on Mackenzie's solution, to dynamically allocate resources. To achieve this, I monitored the pipeline to assess the memory consumption of SynthSR and adjusted the mem_mb parameter from 32000 MB to 10000 MB accordingly. I also modified the thread allocation for this rule, making it dependent on available memory. This ensures that as many subjects as possible can be processed in parallel while avoiding OOM errors. You guys are welcome to test and once I've gotten approval from both of you I'll merge and close this PR!

mackenziesnyder commented:
> > @ataha24 @mackenziesnyder I believe we're in good shape to merge this branch. I've added some additional logic, building on Mackenzie's solution, to dynamically allocate resources. To achieve this, I monitored the pipeline to assess the memory consumption of SynthSR and adjusted the mem_mb parameter from 32000 MB to 10000 MB accordingly. I also modified the thread allocation for this rule, making it dependent on available memory. This ensures that as many subjects as possible can be processed in parallel while avoiding OOM errors. You guys are welcome to test and once I've gotten approval from both of you I'll merge and close this PR!
> 
> That's awesome, thanks Djay! I will test it now

I ran it on 10 subjects using cbs basic and it worked!

Dhananjhay commented:
Sweet, thanks for your excellent work here @mackenziesnyder! Merging it now!
